ai_feedback/helpers/image_extractor.py

```diff
@@ -51,3 +58,234 @@ def extract_images(input_notebook_path: os.PathLike, output_directory: os.PathLi
                     question_context_path = os.path.join(output_directory, question_name, question_context_filename)
                     with open(question_context_path, "w") as txt_file:
                         txt_file.write(question_context_data)
+    return image_paths
+
+
+def extract_qmd_python_chunks_with_context(qmd_path: str) -> List[Dict[str, Any]]:
+    """
+    Extract ONLY Python code chunks from a QMD and annotate each with context from # / ## headings.
+    Supports ```{python ...}, ```python, and ~~~ variants. Skips YAML front matter.
+    """
+    _PY_CHUNK_START = re.compile(
+        r"""^(
+                ```\{python\b[^}]*\}\s*$ |   # ```{python ...}
+                ```python\s*$            |   # ```python
+                ~~~\{python\b[^}]*\}\s*$ |   # ~~~{python ...}
+                ~~~python\s*$                # ~~~python
+            )""",
+        re.IGNORECASE | re.VERBOSE,
+    )
+    _FENCE_END_TICKS = re.compile(r"^```\s*$")
+    _FENCE_END_TILDES = re.compile(r"^~~~\s*$")
+    _H1 = re.compile(r"^#\s+(.*)$")
+    _H2 = re.compile(r"^##\s+(.*)$")
+
+    qp = Path(qmd_path)
+    if not qp.exists():
+        raise FileNotFoundError(f"QMD file not found: {qmd_path}")
+
+    lines = qp.read_text(encoding="utf-8", errors="ignore").splitlines()
+
+    # Skip YAML front matter if present
+    i = 0
+    if lines and re.match(r"^---\s*$", lines[0]):
+        i = 1
+        while i < len(lines) and not re.match(r"^---\s*$", lines[i]):
+            i += 1
+        i = min(i + 1, len(lines))
+
+    current_main = None
+    current_sub = None
+    chunks = []
+    in_py = False
+    cur = []
+    start_line = 0
+    fence_kind = None  # "```" or "~~~"
+
+    for i, raw in enumerate(lines):
+        line = raw.rstrip("\n")
+
+        if not in_py and _PY_CHUNK_START.match(line):
+            in_py = True
+            cur = []
+            start_line = i
+            fence_kind = "~~~" if line.strip().startswith("~~~") else "```"
+            i += 1
```

Comment:
As a result of switching to a for loop, the `i += 1` statements are no longer necessary